### PR TITLE
Additional fixes for IR

### PR DESCRIPTION
**Pull request by reuterbal**
_Note: This sits on top of #44_

I'm filing this separately because it contributes a significant behaviour change for the `SubstituteExpressions` visitor, which I consider to be sensible but that might require extra testing to make sure it doesn't break behaviour in (untested) edge cases.

In a few places, we ran into infinite recursion in the `AttachScopes` visitor (although it's not specific to that one, it's just the first to be encountered within the frontends), more specifically, if a declared variable is being used in one of the definining attributes. Encountered examples include:

* the initializer expression, e.g. `REAL(KIND=JPRB), PARAMETER :: ZEXPLIMIT = LOG(HUGE(ZEXPLIMIT))`
* the allocation dimensions (which are injected as `shape` in the Loki IR), e.g.: `allocate(levels(jscale)%data(size(levels(jscale-1)%data, 1), size(levels(jscale-1)%data, 2)))`
* type bound procedure declarations with an explicit interface of the same name: `procedure (some_proc), deferred :: some_proc`

The cause of the infinite recursion is the fact that we traverse declaration attributes (`type.initial`, `type.shape`, `type.bind_names`) on every symbol use in the `LokiIdentityMapper`. By adding a `recurse_to_declaration_attributes` keyword argument to the visitor methods (defaults to `True` for nodes that are the "authority" on a symbol's type: `VariableDeclaration`, `ProcedureDeclaration`, `Import`), we recurse into these properties only once. Since these are properties that are cached in the symbol table, this is also sufficient (unless transformations rely on the implicit modification of a symbol's declaration when traversing a routine's body, which seems somewhat counterintuitive behaviour).

Conceptually I think this is the right thing to do but the implementation might not be the most elegant...

This also removes quite a bit of traversal overhead in the expression transformer, which shaves off about 10% of the time required to read in cloudsc.F90.

### CODE REVIEW

**Comment by reuterbal** [comment]
I had to update this to account for a specific fparser behaviour (see https://github.com/stfc/fparser/issues/400) which has quite significant impact and for which we may have to watch out.

**Review by mlange05 — CHANGES_REQUESTED** [review]
Ok, first of all, many thanks for figuring this stuff out. I will certainly appreciate 10% of CLOUDSC parsing time reduction. 

I think the behavioural change is fine, but the implementation that exposes details of the outer IR to the inner is not ok, as it goes against the existing visitor pattern and will be hard to remove at a later stage. 

In addition, it might be useful to separate the frontend change relating to fparser from the `SubstituteExpression` change to streamline the merge process a little.

File: loki/expression/mappers.py [hunk 1]
```diff
@@ -520,6 +521,10 @@ def __init__(self, invalidate_source=True):
     def __call__(self, expr, *args, **kwargs):
         if expr is None:
             return None
+        kwargs.setdefault(
+            'recurse_to_declaration_attributes',
+            'current_node' not in kwargs or isinstance(kwargs['current_node'], DECLARATION_NODES)
```
[the remark this answers is not in the document]
Comment (reuterbal) [reply loki/expression/mappers.py:526]: Yes, good point, thanks for pointing that out! I'll try your first proposal, I think it is essentially what I was shooting for but with a much more awkward solution, and I don't remember why I picked that route (it's been so long 😏)

File: loki/expression/mappers.py [hunk 2]
```diff
@@ -557,27 +562,46 @@ def map_variable_symbol(self, expr, *args, **kwargs):
...
             expr.scope.symbol_attrs[expr.name] = expr.type.clone(initial=initial)
 
-        bind_names = self.rec(expr.type.bind_names, *args, **kwargs)
-        if not (bind_names is None or all(new is old for new, old in zip_longest(bind_names, expr.type.bind_names))):
-            # Update symbol table entry for bind_names directly because with a scope attached
-            # it does not affect the outcome of expr.clone
-            expr.scope.symbol_attrs[expr.name] = expr.type.clone(bind_names=as_tuple(bind_names))
+        if kwargs['recurse_to_declaration_attributes']:
+            _kwargs = kwargs.copy()
+            _kwargs['recurse_to_declaration_attributes'] = False
```
[the remark this answers is not in the document]
Comment (reuterbal) [reply loki/expression/mappers.py]: Good point. I have tried to restructure the control flow in a way that makes it clearer what's going on, and add comments.
In essence: only when recurse_to_declaration_attributes is True, we recurse to all the bits that are stored in a symbol's type. But, because Fortran, we don't want this recursion to happen while actually updating these declaration attributes. Reason is that the declared variable can actually appear in its own initializer expression:
```
REAL var = HUGE(var)
```
Or "it" can appear in an allocation:
```
TYPE(SOME_TYPE), ALLOCATABLE :: var(:)
...
DO j=2,n
    ALLOCATE(var(j)%field(size(var(j-1)%field)))
END DO
```
And since we insert the dimensions from allocate statements as `type.shape`, this now becomes also part of the symbol table entries...

[the remark this answers is not in the document]
Comment (mlange05) [reply loki/expression/mappers.py]: Ahhhh, thanks for the explanation. Makes sense and looks much clearer now.

File: loki/expression/mappers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mlange05) [reply loki/expression/mappers.py]: Yes, looks good now. Many thanks for addressing this.
